File: idle/idle/CarPlay/CarPlaySceneDelegate.m
#import "CarPlaySceneDelegate.h"
#import "idle-Swift.h"

@implementation CarPlaySceneDelegate

// MARK: - CPTemplateApplicationSceneDelegate

// iOS 14+ non-navigation variant (no window)
- (void)templateApplicationScene:(CPTemplateApplicationScene *)templateApplicationScene
   didConnectInterfaceController:(CPInterfaceController *)interfaceController {
    self.interfaceController = interfaceController;
    interfaceController.delegate = self;
    [CarPlayBridge didConnectWithInterfaceController:interfaceController delegate:self];
}

// iOS 13+ navigation variant (with window) — implemented so iOS 26 runtime
// finds a matching selector regardless of which variant it checks first.
- (void)templateApplicationScene:(CPTemplateApplicationScene *)templateApplicationScene
   didConnectInterfaceController:(CPInterfaceController *)interfaceController
                        toWindow:(CPWindow *)window {
    self.interfaceController = interfaceController;
    interfaceController.delegate = self;
    [CarPlayBridge didConnectWithInterfaceController:interfaceController delegate:self];
}

- (void)templateApplicationScene:(CPTemplateApplicationScene *)templateApplicationScene
didDisconnectInterfaceController:(CPInterfaceController *)interfaceController {
    [CarPlayBridge didDisconnect];
    self.interfaceController = nil;
}

- (void)templateApplicationScene:(CPTemplateApplicationScene *)templateApplicationScene
didDisconnectInterfaceController:(CPInterfaceController *)interfaceController
                      fromWindow:(CPWindow *)window {
    [CarPlayBridge didDisconnect];
    self.interfaceController = nil;
}

// MARK: - CPInterfaceControllerDelegate

- (void)templateWillAppear:(CPTemplate *)aTemplate animated:(BOOL)animated {}
- (void)templateDidAppear:(CPTemplate *)aTemplate animated:(BOOL)animated {}
- (void)templateWillDisappear:(CPTemplate *)aTemplate animated:(BOOL)animated {}
- (void)templateDidDisappear:(CPTemplate *)aTemplate animated:(BOOL)animated {}

@end

File: idle/idle/CarPlay/CarPlaySceneDelegate.swift
import CarPlay
import UIKit

class CarPlaySceneDelegate: UIResponder, CPTemplateApplicationSceneDelegate {

    var interfaceController: CPInterfaceController?
    private var tabManager: CarPlayTabManager?

    // MARK: - CPTemplateApplicationSceneDelegate
    //
    // These methods MUST be nonisolated. With SWIFT_DEFAULT_ACTOR_ISOLATION = MainActor
    // the class is MainActor-isolated by default, but CPTemplateApplicationSceneDelegate
    // is an Obj-C protocol whose methods are called by the CarPlay runtime from a
    // non-MainActor context. Without nonisolated the Obj-C runtime can't find the
    // selector and the app crashes with "does not implement CarPlay lifecycle methods".

    nonisolated func templateApplicationScene(
        _ templateApplicationScene: CPTemplateApplicationScene,
        didConnect interfaceController: CPInterfaceController
    ) {
        Task { @MainActor in
            self.interfaceController = interfaceController
            interfaceController.delegate = self

            guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else { return }

            let manager = CarPlayTabManager(
                interfaceController: interfaceController,
                serviceRegistry: appDelegate.serviceRegistry,
                queueManager: appDelegate.queueManager,
                playbackEngine: appDelegate.playbackEngine
            )
            self.tabManager = manager
            manager.buildAndSetRoot()

            NotificationCenter.default.post(name: .carPlayDidConnect, object: nil)
        }
    }

    nonisolated func templateApplicationScene(
        _ templateApplicationScene: CPTemplateApplicationScene,
        didDisconnectInterfaceController interfaceController: CPInterfaceController
    ) {
        Task { @MainActor in
            self.interfaceController = nil
            self.tabManager = nil
            NotificationCenter.default.post(name: .carPlayDidDisconnect, object: nil)
        }
    }
}

// MARK: - CPInterfaceControllerDelegate

extension CarPlaySceneDelegate: CPInterfaceControllerDelegate {
    // Also nonisolated — CPInterfaceControllerDelegate is an Obj-C protocol called
    // from a non-MainActor context; same reasoning as the lifecycle methods above.
    nonisolated func templateWillAppear(_ aTemplate: CPTemplate, animated: Bool) {}
    nonisolated func templateDidAppear(_ aTemplate: CPTemplate, animated: Bool) {}
    nonisolated func templateWillDisappear(_ aTemplate: CPTemplate, animated: Bool) {}
    nonisolated func templateDidDisappear(_ aTemplate: CPTemplate, animated: Bool) {}
}

// MARK: - Notification Names

extension Notification.Name {
    static let carPlayDidConnect = Notification.Name("carPlayDidConnect")
    static let carPlayDidDisconnect = Notification.Name("carPlayDidDisconnect")
    static let carPlayRebuildTabs = Notification.Name("carPlayRebuildTabs")
}
